### PR TITLE
Move HPC-GAP specific code (companion to #515) 

### PR DESCRIPTION
**Pull request by markuspf**

### CODE REVIEW

**Comment by markuspf** [comment]
This is not finished yet. The build system needs a tweak, and I am waiting for @rbehrends to comment.

**Comment by markuspf** [comment]
This now builds. Include paths are relative to `src/` which seems to be the only low-pain solution at the moment. Maybe @fingolfin, @ChrisJefferson, @rbehrends can review this as well. I'll continue moving files.

**Comment by fingolfin** [comment]
Thinking about it again, I actually kind of prefer `#include "hpc/aobjects.h"` simply because it also makes it very clear that that line is for HPC. If there are no actual drawbacks to it (I don't recally any specific drawbacks being named), we could just stay with it.

File: src/hpc/tls.h [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+
 #include <stdint.h>
 
-#include "code.h"
+#include "../code.h"
 #include "tlsconfig.h"
-#include "scanner.h"
-#include "gasman.h"
+#include "../scanner.h"
+#include "../gasman.h"
```
Comment (fingolfin) [inline src/hpc/tls.h]: Aha, now I see what you mean with "relative paths". Well, that is slightly ugly. But still not a show stopper -- if we can't fix it now, we certainly will be able to fix it later with the new build system. In any case, this is a minor detail, and not a showstopper.

**Comment by fingolfin** [comment]
Looks good overall (but is marked as DO NOT MERGE!, so I won't ;-)

**Comment by markuspf** [comment]
It's also not building in travis, so I'll investigate.

As to the includes, I have enough `sed` magic around to just edit these things easily now anyway.

**Comment by markuspf** [comment]
The reason it does not build is that `gen/` does not contain the `hpc` directory, but everything is just preprocessed into `gen/`. Willl hack the SConstruct script once more I guess.

**Comment by rbehrends** [comment]
I am not a fan of this solution, as (IMHO) it makes things worse rather than better.

The underlying problem is that we can't use `-Isrc` because then the C compiler gets confused between `<string.h>` and `"string.h"`. This is what we need to fix, not create a jungle of relative include paths that will break terribly the moment we move a file again.

I'd rather use an `hpc-` prefix for files instead of this, but in reality, what we want is to be able to use `-Isrc` properly. Or we could resort to using `-idirafter`. It's not a beacon of portability, but it is supported by all the major compilers, and I'd rather use that than purposely creating difficult to maintain code.

**Comment by ChrisJefferson** [comment]
I never noticed we had a string.h.. we probably should rename that (independant of other issues)

**Comment by fingolfin** [comment]
Aha, I didn't notice the problem with `string.h` either -- good to have that spelled out explicitly.

As to renaming it, note that at least Browse and float seem to explicitly include that head (as `src/string.h`). Using `-idirafter src` seems like a better solution to me, at least for now -- it is supported by gcc, clang and Intel C at least.

**Comment by rbehrends** [comment]
> I never noticed we had a string.h.. we probably should rename that (independant of other issues)

There are not really any other issues (other than teaching SConstruct about the new directory). That's the one and only reason why we can't use `-Isrc`.

**Comment by rbehrends** [comment]
> Using `-idirafter src` seems like a better solution to me, at least for now -- it is supported by gcc, clang and Intel C at least.

While I suggested that, let me also point out that this carries its own risks. It means that if we have a system include file overriding a non-system file, then that can also cause trouble, if we're using `#include <whatever.h>` instead of `#include "whatever.h"`. One example here is actually `gmp.h`, I think.

**Comment by markuspf** [comment]
I updated this PR to reflect the renaming of `string.h` and `string.c` in `master`, but unfortunately the build system picks up the compatibility `string.h` instead of the system's. Hence, I have removed the compatibility header for the time being.

@rbehrends any insights on that?

**Comment by markuspf** [comment]
I will now merge this and continue with the synchronisation of the `master` and `hpcgap-default` branches in terms of where files live, so we can continue with the convergence plan.

**Pull request merged**